### DISCUSSION TITLE
My implementation

### DISCUSSION THREAD

**Discussion by qwasyx0** [spatie/file-system-watcher#20] [General]
Hello, I was able to get the file-system-watcher working on local and production, so I thought I would share, because it is awesome package.
- WatchForFileChangesCommand.php has dispatch(new WatchForFileChangesJob)
- in the handle() method:
```php
$directory = storage_path('app/public/sestavy');
Watch::path($directory)
     ->onFileCreated(function (string $newFilePath) {
        //... whatever you want to do with the file
     })
     ->start();
```
- `sail artisan file:watch` and everything works (signature of my command and I use laravel/sail for local development)

On production
- in /etc/supervisor/conf.d i modified the example configuration:
```bash
[program:watch]
process_name=%(program_name)s_%(process_num)02d
directory=/path/to/project
command=env PATH="/usr/bin" /usr/bin/php /path/to/project/artisan file:watch
startretries=5
autostart=true
autorestart=false
numprocs=1
user=root
redirect_stderr=true
stdout_logfile=/path/to/project/storage/logs/watch.log
stopwaitsecs=3600
```
- then you need to add a queue:worker for the command to run
```bash
[program:queue-worker-crm]
process_name=%(program_name)s_%(process_num)02d
command=php /path/to/project/artisan queue:work --tries=3
autostart=true
autorestart=true
user=root
numprocs=1
redirect_stderr=true
stdout_logfile=/path/to/project/storage/logs/supervisord.log
```
- you must have only one worker, otherwise every worker will execute the command (atleast from my knowledge)

run commands:
`supervisorctl reread`
`supervisorctl update`
`service supervisor restart`
- when you run supervisorctl, you should see 2 new processess:

![Snímek obrazovky 2023-02-25 v 19 05 48](https://user-images.githubusercontent.com/43924784/221372717-c59c9c8b-cef6-4d1c-81c5-b3eb61fe2544.png)
- watch process will be run by queue:worker, so it's fine its in FATAL for now

In the project
- you need to have node_modules installed, otherwise you will get error `Could not start watcher. Make sure you have required chokidar...`
- try to save file in storage
- in storage/logs/supervisor.log you should see every minute a new call of App\Jobs\WatchForFileChangesJob

![Snímek obrazovky 2023-02-25 v 19 09 42](https://user-images.githubusercontent.com/43924784/221372850-84e9a203-8bda-46cb-a860-849d525504d8.png)
- if there are any errors in the watcher, they will be logged in storage/logs/watch.log

Hope it's helpful to someone, I might modify it later with new information. Tips for improvement appreciated!

After few hours, I got an error `Error: ENOSPC: System limit for number of file watchers reached, watch`
- you need to increase number of watchers
- `cat /proc/sys/fs/inotify/max_user_watches` returned 256
- run `echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf` to increase to 524288
- `sudo sysctl -p` have the linux use the new limit
- `cat /proc/sys/fs/inotify/max_user_watches` should return 524288
- [Reference](https://klequis.io/error-enospc-system-limit-for-number-of-file-watchers-reached/)
- maybe it will be a good practice to restart the supervisor every night to reset all watchers, I will test it

**Comment by henrytorres9** [comment]
Hi, I am new to programming, I have added the repository in my laravel project but I don't know exactly how to do the test you mention locally.

I am using laravel sail anyway, could you guide me how to do the test, would there be a way for you to make a video?
